Device/Device.cs

```diff
@@ -68,36 +60,44 @@ public async Task ConnectDevice()
             mqttClient.UseDisconnectedHandler(new MqttClientDisconnectedHandlerDelegate(e => Disconnected(e, options)));
         }
 
-        public async Task SendCloudToDeviceMessageAsync(string payload)
-        {
-            var topicD2C = $"devices/{deviceId}/messages/events/";
-
-            Console.WriteLine($"Topic:{topicD2C} Payload:{payload}");
-
+        public MqttApplicationMessage ConstructMessage(string topic, string payload, bool retainFlag = false)
+        {
+            /*var payloadJObject = new JObject
+                {
+                    { "OfficeTemperature", payload + DateTime.UtcNow.Millisecond.ToString() }
+                };
+
+            payload = JsonConvert.SerializeObject(payloadJObject);*/
+            Console.WriteLine($"Topic:{topic} Payload:{payload}");
+
             var message = new MqttApplicationMessageBuilder()
-                .WithTopic(topicD2C)
+                .WithTopic(topic)
                 .WithPayload(payload)
+                .WithRetainFlag(retainFlag)
                 .WithAtLeastOnceQoS()
-                .Build();
+                .Build();
+
+            return message;
+        }
+
+
+        public async Task SendDeviceToCloudMessageAsync(string payload, bool retainFlag = false)
+        {
+            var topicD2C = $"devices/{deviceId}/messages/events/$.ct=application%2Fjson&$.ce=utf-8";
+            var message = ConstructMessage(topicD2C, payload, retainFlag);
 
             Console.WriteLine("PublishAsync start");
             await mqttClient.PublishAsync(message, CancellationToken.None);
             Console.WriteLine("PublishAsync finish");
         }
 
-        public async Task SendEventsAsync()
+        public async Task SendD2CMessagesInALoopAsync() // TODO: delete if no longer used
         {
+            var payload = 0;
             while (true)
             {
-                var payloadJObject = new JObject();
-
-                payloadJObject.Add("OfficeTemperature", "22." + DateTime.UtcNow.Millisecond.ToString());

```

Comment:
Why did you remove the payload creation here?

Reply:
To be able to control the content of the payload from only what I pass as a parameter. And also to be able to verify the value in the unit test without serialising/deserialising.

Reply:
If you want to do that why not just use the send message a single message function? I created this one to see constant stream of semi random messages sent to Iot hub.